proposals/span-safety.md

```diff
@@ -0,0 +1,104 @@
+2. Some implementations of `Span<T>` literally contain a managed pointer in one of its fields. Managed pointers are not supported as fields of heap objects and code that manages to put a managed pointer on the GC heap typically crashes at JIT time.
+3. It is permitted for a `Span<T>` to refer to data in the local stack frame - individual local variables or `stackalloc`-ed arrays. A scenario when an instance of a `Span<T>` outlives the referred data would lead to undefined behavior, including type-safety violations and heap corruptions.
+
+All the above problems would be alleviated if instances of `Span<T>` would be allowed as stack data types only. In fact for the `#3` we need a stronger guarantee. Details on that to follow below.
+
+An additional problem arises due to composition. It would be generally desirable to build more complex data types that would embed `Span<T>` instances. Such composite types would have to be structs and would share all the hazards and requirements of `Span<T>`. As a result the safety rules described here should be viewed as applicable to the whole range of **_ref-like types_**
+
+At the minimum the restrictions described below would need to apply to `Span<T>`.
+
+## `ref-like` types must be stack-only. ##
+
+C# compiler already has a concept of a restricted types that covers special platform types such as `TypedReference`. In some cases those types are referred as ref-like types as well. We should probably use some different term to refer to `TypedReference` and similar types - like `restricted types` to avoid confusion.
+
+In this document the "ref like" types include only `Span<T>` and related types. And the ref-like variables mean variables of "ref-like" types.  
+
+In order to force ref-like variables to be stack only, we need the following restrictions: 
+
+- ref-like type cannot be a type of an array element
+- ref-like type cannot be used as a generic type argument
+- ref-like variable cannot be boxed
+- ref-like type cannot be a field of ordinary not ref-like type
+- indirect restrictions, such as disallowed use of ref-like types in async methods, which are really a result of disallowing ref-like typed fields.
+
+Note:
+
+- it is ok to allow returning ref-like values from methods (additional rules apply when referring to local data)
+- it is ok for a ref-like type be a field of another struct as long as the struct itself is ref-like.
+
+
+## ref-like variables can be passed as `in` parameters ##
+
+- it is ok to pass a ref-like variable by reference to a call as long as the reference is **_readonly_** . 
+The reason is that exposing ref-like variables via writeable references make it nearly impossible to reason about their safety.
```

Comment:
Unsafe code is a tough subject. Some unsafe code has well-defined behavior even if not safe.  Your example may actually run just fine depending on implementation. 
C# as a language generally does not have a lot of guarantees about unsafe code.

I'd say the sample is legal, but its behavior is implementation specific and thus undefined.